packages/babel-core/README.md

```diff
@@ -123,6 +123,78 @@ const parsedAst = babylon.parse(sourceCode, { allowReturnOutsideFunction: true }
+* It can be safely passed back to Babel. Fields like `babelrc` have been set to 
+  false so that later calls to Babel will not make a second attempt to load
+  config files.
+
+`Plugin` instances aren't meant to be manipulated directly, but often
+callers will serialize this `opts` to JSON to use it as a cache key representing
+the options Babel has received. Caching on this isn't 100% guaranteed to
+invalidate properly, but it is the best we have at the moment.
+
+
+### babel.loadPartialConfig([options?](#options): Object): PartialConfig
+
+To allow systems to easily manipulate and validate a user's config, this function
+resolves the plugins and presets and proceeds no further. The expectation is
+that callers will take the config's `.options`, manipulate it as then see fit
+and pass it back to Babel again.
+
+* `babelrc: string | void` - The path of the `.babelrc` file, if there was one.
+* `babelignore: string | void` - The path of the `.babelignore` file, if there was one.
+* `options: ValidatedOptions` - The partially resolved options, which can be manipulated and passed back to Babel again.
+  * `plugins: Array<ConfigItem>` - See below.
+  * `presets: Array<ConfigItem>` - See below.
+  * It can be safely passed back to Babel. Fields like `babelrc` have been set
+    to false so that later calls to Babel will not make a second attempt to 
+    load config files.
+
+[`ConfigItem`](#configitem-type) instances expose properties to introspect the values, but each
+item should be treated as immutable. If changes are desired, the item should be
+removed from the list and replaced with either a normal Babel config value, or
+with a replacement item created by `babel.createConfigItem`. See that
+function for information about `ConfigItem` fields.
+
+
+### babel.createConfigItem(value: string | {} | Function, options?: {}, { dirname?: string, name?: string, type?: "preset" | "plugin" }): ConfigItem
+
+Allows build tooling to create and cache config items up front. If this function
+is called multiple times for a given plugin, Babel will call the plugin's function itself
+multiple times. If you have a clear set of expected plugins and presets to
+inject, pre-constructing the config items would be recommended.
+
+
+### `ConfigItem` type
+
+Each `ConfigItem` exposes all of the information Babel knows. The fields are:
+
+* `value: {} | Function` - The resolved value of the plugin.
+* `options: {} | void` - The options object passed to the plugin.
+* `dirname: string` - The path that the options are relative to.
+* `name: string | void` - The name that the user gave the plugin instance, e.g. `plugins: [ ['env', {}, 'my-env'] ]` 
+* `file: Object | void` - Information about the plugin's file, if Babel knows it.
+  * `request: string` - The file that the user requested, e.g. `"@babel/env"`
+  * `resolved: string` - The full path of the resolved file, e.g. `"/tmp/node_modules/@babel/preset-env/lib/index.js"`
```

Comment:
This is awesome, so in theory babel-loader can add every single plugin/preset as dependency (if coming from a file) so that webpack watch will trigger a change when one of them changes.

Reply:
Potentially, but since the presets aren't flattened, there are other plugins that aren't included here, so it wouldn't cover everything. Also since they are CommonJS, they'll still be cached by the module loader unless something also cleared that. I mostly exposed this since we already knew it and it seemed helpful for users. I'm figuring this will be good because they we can dump out a user's config in a human-readable format mostly.